File: day_4/day_4_1.rb
# frozen_string_literal: true

class CampCleanup
  def initialize(filename)
    assignments = parse_assignments(filename)
    
    puts "Of #{assignments.size} assignments, #{count_redundant_assignments(assignments)} are redundant."
  end

  def parse_assignments(filename)
    File.read(filename).split("\n").map do |line|
      line.split(",").map do |assignment|
        min, max = assignment.split("-").map(&:to_i)
        (min..max)
      end
    end
  end

  def count_redundant_assignments(assignments)
    assignments.select do |left, right|
      redundant?(left, right)
    end.size
  end

  def redundant?(left, right)
    smaller, larger = sort_by_containment(left, right)
    smaller.min >= larger.min && smaller.max <= larger.max
  end

  #####
  #
  # The version below is arguably a better implementation, but take a look at the Rust version.
  # It felt like it was worth it to do it this way to get an opportunity to talk about one of the
  # trickier aspects of Rust's syntax, if the solution were structured as above instead of like this.
  # So....leaving this for reference, but keeping the above solution for the writeup.
  #
  #####
  #
  # def redundant?(left, right)
  #   if left.min <= right.min && left.max >= right.max
  #     true
  #   elsif left.min >= right.min && left.max <= right.max
  #     true
  #   else
  #     false  
  #   end
  # end
  #
  #####

  def sort_by_containment(left, right)
    if left.min <= right.min && left.max >= right.max
      [right, left]
    else
      [left, right]
    end
  end
end

CampCleanup.new(ARGV[0] || 'input.txt')

File: day_4/day_4_1.rs
use std::env;

struct CampCleanup {
    assignments: Vec<Vec<Assignment>>,
}

#[derive(Debug)]
struct Assignment {
    start: i32,
    end: i32,
}

impl CampCleanup {
    fn new(filename: &str) -> CampCleanup {
        let assignments = Self::parse_assignments(filename);

        //CampCleanup { assignments }
        Self { assignments }
    }

    fn run(&self) {
        println!(
            "Of {} assignments, {} are redundant.",
            self.assignments.len(),
            self.count_redundant_assignments(&self.assignments)
        );
    }

    fn parse_assignments(filename: &str) -> Vec<Vec<Assignment>> {
        let text = std::fs::read_to_string(filename).unwrap();
        text.lines()
            .map(|line| {
                line.split(",")
                    .map(|assignment| {
                        let mut parts = assignment.split("-");
                        let min = parts.next().unwrap().parse().unwrap();
                        let max = parts.next().unwrap().parse().unwrap();
                        Assignment {
                            start: min,
                            end: max,
                        }
                    })
                    .collect::<Vec<_>>()
            })
            .collect::<Vec<_>>()
    }

    fn count_redundant_assignments(&self, assignments: &Vec<Vec<Assignment>>) -> usize {
        assignments
            .iter()
            .filter(|assignment| self.is_redundant(&assignment[0], &assignment[1]))
            .count()
    }

    fn is_redundant(&self, left: &Assignment, right: &Assignment) -> bool {
        let (smaller, larger) = self.sort_by_containment(left, right);
        smaller.start >= larger.start && smaller.end <= larger.end
    }

    fn sort_by_containment<'a>(
        &'a self,
        left: &'a Assignment,
        right: &'a Assignment,
    ) -> (&Assignment, &Assignment) {
        if left.start <= right.start && left.end >= right.end {
            (right, left)
        } else {
            (left, right)
        }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut filename = "input.txt";
    if let Some(arg) = args.get(1) {
        filename = arg;
    }

    CampCleanup::new(filename).run();
}
